control_plane/attachment_service/src/heartbeater.rs

```diff
@@ -0,0 +1,203 @@
+use futures::{stream::FuturesUnordered, StreamExt};
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio_util::sync::CancellationToken;
+
+use pageserver_api::models::PageserverUtilization;
+
+use thiserror::Error;
+use utils::id::NodeId;
+
+use crate::node::Node;
+
+pub struct Heartbeater {
+    receiver: tokio::sync::mpsc::UnboundedReceiver<HeartbeatRequest>,
+    cancel: CancellationToken,
+
+    state: Arc<HashMap<NodeId, PageserverState>>,
+
+    max_unavailable_interval: Duration,
+    jwt_token: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+pub enum PageserverState {
+    Available {
+        last_seen_at: Instant,
+        utilization: PageserverUtilization,
+    },
+    Offline,
+}
+
+#[derive(Debug)]
+pub struct AvailablityDeltas(pub Vec<(NodeId, PageserverState)>);
+
+#[derive(Debug, Error)]
+pub enum HeartbeaterError {
+    #[error("Cancelled")]
+    Cancel,
+}
+
+pub struct HeartbeatRequest {
+    pageservers: Arc<HashMap<NodeId, Node>>,
+    reply: tokio::sync::oneshot::Sender<Result<AvailablityDeltas, HeartbeaterError>>,
+}
+pub struct HeartbeaterHandler {
+    sender: tokio::sync::mpsc::UnboundedSender<HeartbeatRequest>,
+}
+
+impl HeartbeaterHandler {
```

Comment:
nit: the "handler" part of the name feels redundant, imo "heartbeater" or "heartbeats" would be fine

Reply:
Fixed in https://github.com/neondatabase/neon/pull/7092/commits/73db76c228c74f4b5a4660cec02b8aebb3921aaf and https://github.com/neondatabase/neon/pull/7092/commits/cd6fbceff2875645c5911198227e432ff2a5ceb1